### PR TITLE
COMP: 'Where' keyword completion

### PR DESCRIPTION
**Pull request by Undin**

### CODE REVIEW

File: src/main/grammars/RustParser.bnf [hunk 1]
```diff
@@ -562,6 +562,7 @@ private UnitStructTail ::= WhereClause? ';'
 private BlockStructTail ::= WhereClause? BlockFields
 
 EnumItem ::= AttrsAndVis enum identifier TypeParameterList? WhereClause? EnumBody {
+  pin = "enum"
```
Comment (matklad) [inline src/main/grammars/RustParser.bnf]: 👍

File: src/main/kotlin/org/rust/lang/core/completion/RsKeywordCompletionContributor.kt [hunk 2]
```diff
@@ -96,24 +95,48 @@ class RsKeywordCompletionContributor : CompletionContributor(), DumbAware {
...
+            .andNot(psiElement().withParent(RsTraitOrImpl::class.java))
+
+        val trait = psiElement<RsTraitItem>()
+            .withLastChildSkipping(RsPsiPattern.error, or(psiElement(IDENTIFIER), typeParameters))
+
+        val impl = psiElement<RsImplItem>()
+            .withLastChildSkipping(RsPsiPattern.error, psiElement<RsTypeReference>())
+
+        return psiElement()
+            .withPrevSiblingSkipping(RsPsiPattern.whitespace, or(function, struct, enum, typeAlias, trait, impl))
```
Comment (matklad) [inline src/main/kotlin/org/rust/lang/core/completion/RsKeywordCompletionContributor.kt]: Ideally, we should use parser to provide keyword completion, like Kotlin does: https://github.com/JetBrains/kotlin/blob/master/idea/idea-completion/src/org/jetbrains/kotlin/idea/completion/KeywordCompletion.kt#L199.

But pattern based approach should work quite well in practice!

**Comment by matklad** [comment]
bors r+

Thanks!

**Comment by matklad** [comment]
Looks like this needs some rebasing now because of https://github.com/intellij-rust/intellij-rust/commit/85cc8ce21185e61c62d4f398606ccb8fee9ecb63

**Comment by Undin** [comment]
Done

**Pull request merged**